### PR TITLE
chore(Bonus Pagamenti Digitali): [IAC-140] Starts opt-in payment methods workunit

### PR DESCRIPTION
**Pull request by debiff**
## Short description
This PR manages the logic that starts the opt-in payment methods workunit.

## List of changes proposed in this pull request
- Added the `optInShouldShowChoiceHandler` saga
- Managed the loading in the `WalletHomeScreen` using the new `ShowOptInChoice` state
- Added the `bpdActivationUiReducer` reducer in the `bpdActivationReducer`

https://user-images.githubusercontent.com/11773070/155297770-d97577e4-9e5d-4f6b-9b8b-7fb94228cf25.mp4

## How to test
Run the tests suite.

To test manually:
- Activate the local `bpdOptInPaymentMethodsEnabled` feature flag and [the remote one](https://github.com/pagopa/io-dev-api-server/blob/43be5e67da75de80c6a91ed1c0a0e5c1dabb1585/src/payloads/backend.ts#L214)
- Enroll the user in the bpd program.
- Play with https://github.com/pagopa/io-dev-api-server/blob/43be5e67da75de80c6a91ed1c0a0e5c1dabb1585/src/routers/features/bdp/index.ts#L49 or https://github.com/pagopa/io-dev-api-server/blob/43be5e67da75de80c6a91ed1c0a0e5c1dabb1585/src/routers/features/bdp/index.ts#L80 to explore the different cases.

### CODE REVIEW

File: ts/features/bonus/bpd/saga/orchestration/optInPaymentMethods/optInShouldShowChoiceHandler.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      )
+    );
+    return;
+  }
+
+  // If the user is not enrolled in the bpd program returns with value false
+  if (!bpdEnabled.value) {
+    yield* put(optInPaymentMethodsShowChoice.success(bpdEnabled.value));
+    return;
+  }
```
Comment (fabriziofff) [inline ts/features/bonus/bpd/saga/orchestration/optInPaymentMethods/optInShouldShowChoiceHandler.ts:67]: At the moment, from a store point of view, we don't distinguish between a user who has never been enrolled to cashback (response `404`) and a user who has been registered but has unsubscribed (`200` with body `enabled: false`). We map both these cases to `enabled: false`.
If I understand correctly, with the current implementation, we never ask someone who has unsubscribed from cashback if he wants to delete their payment methods. It's correct?

**Review by fabriziofff — CHANGES_REQUESTED** [review]

File: ts/features/bonus/bpd/components/walletCardContainer/BpdCardsInWalletComponent.tsx [hunk 2]
```diff
@@ -42,19 +40,14 @@ const BpdCardListMemo = React.memo(
...
-const refreshTime = 300000 as Millisecond;
-
 /**
  * Render the bpd card in the wallet
  * //TODO: handle error, render only if some
  * @constructor
  */
-const BpdCardsInWalletContainer = (props: Props) => {
-  // If the user does "pull to refresh", this timer is ignored and the refresh is forced
-  useActionOnFocus(props.load, refreshTime);
```
[the remark this answers is not in the document]
Comment (debiff) [reply ts/features/bonus/bpd/components/walletCardContainer/BpdCardsInWalletComponent.tsx:55]: Sure, removed in [a7e7fe2](https://github.com/pagopa/io-app/pull/3785/commits/a7e7fe28ac3bf7b00ed9dc70f50cf384ec22f53f)

File: ts/features/bonus/bpd/saga/orchestration/optInPaymentMethods/optInShouldShowChoiceHandler.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+ * - check if the loading of the payment method succeed
+ * - if succeed start the saga
+ */
+export function* optInShouldShowChoiceHandler(): Generator<
+  ReduxSagaEffect,
+  void,
+  any
+> {
+  // Load the information about the participation of the user to the bpd program
+  yield* put(bpdAllData.request());
```
[the remark this answers is not in the document]
Comment (debiff) [reply ts/features/bonus/bpd/saga/orchestration/optInPaymentMethods/optInShouldShowChoiceHandler.ts:38]: I choose to remove the `bpdAllData.request` from the `didMount` of the `WalletHome` screen.
Removed in [6220ee5](https://github.com/pagopa/io-app/pull/3785/commits/6220ee5da597bdabe4e978174ee5bb20d102dcc7)

File: ts/features/bonus/bpd/saga/orchestration/optInPaymentMethods/optInShouldShowChoiceHandler.ts (hunk 1 above)
Comment (debiff) [reply ts/features/bonus/bpd/saga/orchestration/optInPaymentMethods/optInShouldShowChoiceHandler.ts]: You're right. I initially misunderstood the semantics of the response.
Refactored in [d44d979](https://github.com/pagopa/io-app/pull/3785/commits/d44d97909f9b232ee144ce0d984842447b60e643), [31abdcd](https://github.com/pagopa/io-app/pull/3785/commits/31abdcd8eb6c2aa6efd8d3a2779a46f997c949e3)

File: ts/screens/wallet/WalletHomeScreen.tsx [hunk 4]
```diff
@@ -230,7 +246,36 @@ class WalletHomeScreen extends React.PureComponent<Props, State> {
...
+      this.props.runOptInShouldShowChoiceHandler();
+      this.props.showModal(
+        <LoadingSpinnerOverlay
+          isLoading={true}
+          loadingCaption={I18n.t("global.remoteStates.loading")}
+          loadingOpacity={1}
+        />
+      );
+    }
+
```
[the remark this answers is not in the document]
Comment (debiff) [reply ts/screens/wallet/WalletHomeScreen.tsx]: Thank you. Refactored in [4bdc87f](https://github.com/pagopa/io-app/pull/3785/commits/4bdc87f56db2ac34b51696606b479be2be92305a)

**Review by fabriziofff — CHANGES_REQUESTED** [review]

**Review by fabriziofff — APPROVED** [review]

**Pull request merged**